tests/copp/test_copp.py

```diff
@@ -0,0 +1,201 @@
+        the RPC image and install it before the test runs.
+
+        These test cases limit the PPS of all trap groups to 600. This is done to ensure
+        that the PTF can send traffic fast enough to trigger the policer. In order to validate
+        higher rate limits, a physical traffic generator is needed, which is beyond the scope
+        of these test cases.
+
+    Parameters:
+        --nn_target_port <port> (int): Which port you want the test to send traffic
+            to. Default is 3.
+
+            Note that this is not the same as the interface name. For example, Ethernet12
+            may not be the 12th port in your system depending on the HWSKU under test.
+
+        --pkt_tx_count <n> (int): How many packets to send during each individual test case.
+            Default is 100000.
+
+        --swap_syncd: Used to install the RPC syncd image before running the tests. Default
+            is disabled.
+"""
+
+import time
+from collections import namedtuple
+
+import pytest
+from copp import copp_utils
+from ptf_runner import ptf_runner
+from common.system_utils import docker
+from common.broadcom_data import is_broadcom_device
+
+_COPPTestParameters = namedtuple("_COPPTestParameters",
+                                 ["nn_target_port",
+                                  "pkt_tx_count",
+                                  "swap_syncd",
+                                  "topo"])
+_SUPPORTED_TOPOS = ["ptf32", "ptf64", "t1", "t1-lag"]
+_TEST_RATE_LIMIT = 600
+
+class TestCOPP(object):
+    """
+        Tests basic COPP functionality in SONiC.
+    """
+
+    @pytest.mark.parametrize("protocol", ["ARP",
+                                          "IP2ME",
+                                          "SNMP",
+                                          "SSH"])
+    def test_policer(self, protocol, duthost, ptfhost, _copp_testbed):
+        """
+            Validates that rate-limited COPP groups work as expected.
+
+            Checks that the policer enforces the rate limit for protocols
+            that have a set rate limit.
+        """
+
+        if protocol == "ARP" \
+                and is_broadcom_device(duthost) \
+                and "201811" not in duthost.get_version():
+            pytest.xfail("ARP policy disabled on BRCM devices due to SAI bug")
+
+        if protocol in ["IP2ME", "SNMP", "SSH"] and _copp_testbed.topo == "t1-lag":
+            pytest.xfail("Packets not received due to faulty DIP, see #1171")
+
+        _copp_runner(duthost,
+                     ptfhost,
+                     protocol,
+                     _copp_testbed)
+
+    @pytest.mark.parametrize("protocol", ["BGP",
+                                          "DHCP",
+                                          "LACP",
+                                          "LLDP",
+                                          "UDLD"])
+    def test_no_policer(self, protocol, duthost, ptfhost, _copp_testbed):
```

Comment:
What is the reason for `test_policer` and `test_no_policer` wrappers if both of them just run ptftest? 
Each `copp_tests.{protocol}Test` directly inherited from `policer` or `no_policer` test inside ptftest.

Reply:
I just added it so the type of test would appear in the output. I found it helpful when I was debugging, but I can update the comments or revert it back to one set of tests if this is confusing.

Reply:
No need to revert, I think. For me, it looks like a good skeleton for ptfadapter implementation